src/raster/mdarray.rs

```diff
@@ -0,0 +1,331 @@
+    pub fn read_as<T: Copy + GdalType>(
+        &self,
+        array_start_index: Vec<u64>,
+        count: Vec<usize>,
+    ) -> Result<Vec<T>> {
+        let pixels: usize = count.iter().product();
+        let mut data: Vec<T> = Vec::with_capacity(pixels);
+
+        // Safety: the read_into_slice line below writes
+        // exactly pixel elements into the slice, before we
+        // read from this slice. This paradigm is suggested
+        // in the rust std docs
+        // (https://doc.rust-lang.org/std/vec/struct.Vec.html#examples-18)
+        unsafe {
+            data.set_len(pixels);
+        };
+        self.read_into_slice(&mut data, array_start_index, count)?;
+
+        Ok(data)
+    }
+
+    #[cfg(feature = "ndarray")]
+    /// Read a 'Array2<T>' from this band. T implements 'GdalType'.
+    ///
+    /// # Arguments
+    /// * window - the window position from top left
+    /// * window_size - the window size (GDAL will interpolate data if window_size != array_size)
+    /// * array_size - the desired size of the 'Array'
+    /// * e_resample_alg - the resample algorithm used for the interpolation
+    /// # Docs
+    /// The Matrix shape is (rows, cols) and raster shape is (cols in x-axis, rows in y-axis).
+    pub fn read_as_array<T: Copy + GdalType + Debug>(
+        &self,
+        array_start_index: Vec<u64>,
+        count: Vec<usize>,
+        array_size: Vec<usize>,
+    ) -> Result<ArrayD<T>> {
+        let data = self.read_as::<T>(array_start_index, count)?;
+        // Matrix shape is (rows, cols) and raster shape is (cols in x-axis, rows in y-axis)
+
+        let dim: IxDyn = IxDyn(&array_size);
+        Ok(ArrayD::from_shape_vec(dim, data)?)
+    }
+}
+
+/// Represent a mdarray in a dataset
+///
+/// This object carries the lifetime of the dataset that
+/// contains it. This is necessary to prevent the dataset
+/// from being dropped before the group.
+#[derive(Debug)]
+pub struct Group<'a> {
+    c_group: GDALGroupH,
+    dataset: &'a Dataset,
+}
+
+impl<'a> Group<'a> {
+    /// Create a Group from a wrapped C pointer
+    ///
+    /// # Safety
+    /// This method operates on a raw C pointer
+    pub unsafe fn from_c_group(dataset: &'a Dataset, c_group: GDALGroupH) -> Self {
+        Group { c_group, dataset }
+    }
+    pub fn name(&self) -> String {
+        _string(unsafe { GDALGroupGetName(self.c_group) })
+    }
+
+    pub fn array_names(&self, options: CslStringList) -> Vec<String> {
+        let options = options.as_ptr();
+        let c_array_names = unsafe { GDALGroupGetMDArrayNames(self.c_group, options) };
+        _string_array(c_array_names)
+    }
+
+    pub fn md_array(&self, name: String, options: CslStringList) -> MDArray {
+        let name = CString::new(name).unwrap();
+        let c_mdarray =
+            unsafe { GDALGroupOpenMDArray(self.c_group, name.as_ptr(), options.as_ptr()) };
```

Comment:
> Returns    the array names, to be freed with CSLDestroy()

Do we do this somewhere?

I guess @jdroenner used valgrind at some point. Maybe we can do this here. I also struggle to get these C-calls and frees right, so maybe this helps spotting errors.